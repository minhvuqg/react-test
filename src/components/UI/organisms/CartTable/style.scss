.cart-table {
  font-family: Arial, Helvetica, sans-serif;
  border-collapse: collapse;
  width: 100%;
  min-width: 800px;
  border: 1px solid #ddd;

  thead {
    border: 1px solid #ddd;

    th {
      padding-top: 12px;
      padding-bottom: 12px;
      text-align: left;
      background-color: #04aa6d;
      color: white;
      border-bottom: 1px solid #ddd;
    }
  }

  td,
  td {
    padding: 8px 16px;
  }

  tr:nth-child(even) {
    background-color: #f2f2f2;
  }

  tr:hover {
    background-color: #ddd;
  }
}
