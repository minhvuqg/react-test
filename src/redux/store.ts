import { combineReducers } from 'redux'
import { configureStore, getDefaultMiddleware } from '@reduxjs/toolkit'

import productSlice from './reducers/productSlice'

const customizedMiddleware: any = getDefaultMiddleware({
  serializableCheck: false
})

const persistedReducer: any = combineReducers({
  cart: productSlice
})

export default configureStore({
  reducer: persistedReducer,
  middleware: customizedMiddleware
})
