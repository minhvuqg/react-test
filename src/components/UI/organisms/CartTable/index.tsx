import { FC, useMemo, useCallback } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import { ProductQuantity } from 'components/UI/molecules/ProductQuantity'
import { convertToCurrency } from 'helpers/string'
import { addToCart } from 'redux/reducers/productSlice'

import { IProductInCart, StoreType } from 'redux/type.d'
import { IProps } from './type.d'

import './style.scss'

const CartTable: FC<IProps> = (): JSX.Element => {
  const productsInCart: IProductInCart[] = useSelector(
    (state: StoreType): IProductInCart[] => state.cart.list
  )
  const dispatch: any = useDispatch()

  const total: string = useMemo((): string => {
    let t: number = 0

    productsInCart.forEach((product: IProductInCart): void => {
      t += product.price * product.quantity
    })

    return convertToCurrency(t)
  }, [productsInCart])

  const incrementQuantity: any = useCallback((product: IProductInCart): void => {
    dispatch(addToCart({ ...product, quantity: product.quantity + 1 }))
  }, [dispatch])

  const decrementQuantity: any = useCallback((product: IProductInCart): void => {
    const { quantity } = product
    if (quantity > 1) {
      dispatch(addToCart({ ...product, quantity: quantity - 1 }))
    }
  }, [dispatch])

  return (
    <table className="cart-table">
      <thead>
        <tr>
          <td>Product name</td>
          <td className="-text-center">Quantity</td>
          <td className="-text-right">Price</td>
        </tr>
      </thead>

      <tbody>
        {productsInCart.map(
          (product: IProductInCart, index: number): JSX.Element => {
            return (
              <tr key={`product_${index}`}>
                <td>{product.name}</td>
                <td className="-centered">
                  <ProductQuantity
                    quantity={product.quantity}
                    incrementQuantity={(): void => incrementQuantity(product)}
                    decrementQuantity={(): void => decrementQuantity(product)}
                  />
                </td>
                <td className="-text-right">
                  {convertToCurrency(product.price)}
                </td>
              </tr>
            )
          }
        )}
        <tr className="total-row">
          <td colSpan={3} className="-text-right">
            Total: {total}
          </td>
        </tr>
      </tbody>
    </table>
  )
}

export default CartTable
