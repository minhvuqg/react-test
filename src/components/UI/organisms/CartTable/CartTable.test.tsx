import TestRenderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import CartTable from '../CartTable'
import store from 'redux/store'

it('render cart correctly', () => {
  const cartTableEl = TestRenderer.create(
    < Provider store={store} >
      <CartTable />
    </Provider >).toJSON()
  expect(cartTableEl).toMatchSnapshot()
})
