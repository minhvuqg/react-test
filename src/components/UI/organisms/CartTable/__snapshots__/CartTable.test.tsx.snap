// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render cart correctly 1`] = `
<table
  className="cart-table"
>
  <thead>
    <tr>
      <td>
        Product name
      </td>
      <td
        className="-text-center"
      >
        Quantity
      </td>
      <td
        className="-text-right"
      >
        Price
      </td>
    </tr>
  </thead>
  <tbody>
    <tr
      className="total-row"
    >
      <td
        className="-text-right"
        colSpan={3}
      >
        Total: 
        $0.00
      </td>
    </tr>
  </tbody>
</table>
`;
